### PR TITLE
E2E tests now run in multiple clouds in addition to KIND

### PR DESCRIPTION
**Pull request by dsu-igeek**
Split plug-in provider into cloud provider/object provider
Moved velero install/uninstall for tests into velero_utils
Added dump of velero describe and logs on error

Signed-off-by: Dave Smith-Uchida <dsmithuchida@vmware.com>

### CODE REVIEW

File: test/e2e/velero_utils.go
Comment (zubron) [inline test/e2e/velero_utils.go:7]: I think you'll need to run `make update` to sort these imports.

File: test/e2e/velero_utils.go
Comment (zubron) [inline test/e2e/velero_utils.go:227]: Suggestion: Would it make sense to do this validation upfront before running the tests? e.g adding a flag validation step in `TestE2e`

File: test/e2e/velero_utils.go
Comment (zubron) [inline test/e2e/velero_utils.go:199]: This is fine for now, but I think in the long run it would be good to have these commands take configurable `io.Writer`s or write the contents of `Stdout` and `Stderr` to buffers where the caller can decide where to redirect them as we do in the [`exec` util package](https://github.com/vmware-tanzu/velero/blob/main/pkg/util/exec/exec.go#L27-L54).

File: test/e2e/e2e_suite_test.go
Comment (zubron) [inline test/e2e/e2e_suite_test.go:26]: If this is configurable, then the other utility functions in `velero_utils.go` (those not added in this PR) will need to be updated as they all assume the default velero namespace now.

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go:199]: Yes - actually, what I would like to do is fold this stuff into "velero debug" and call "velero debug" after each run and store things away in whatever format we use there.

File: test/e2e/e2e_suite_test.go
Comment (dsu-igeek) [reply test/e2e/e2e_suite_test.go:26]: Good catch - I will update those

File: test/e2e/Makefile
Comment (dsu-igeek) [reply test/e2e/Makefile]: Sure, I updated the message

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go]: Yup, done

File: test/e2e/e2e_suite_test.go
Comment (dsu-igeek) [reply test/e2e/e2e_suite_test.go:26]: OK, those are updated now

**Review by zubron — COMMENTED** [review]
Thanks for making the changes, @dsu-igeek. There is a [codespell error](https://github.com/vmware-tanzu/velero/pull/3286/checks?check_run_id=1767048476) blocking it from merging but otherwise it's good. Once the CI checks are green, I'll approve (pretty sure GitHub will dismiss my review once you make a change).

**Review by nrb — COMMENTED** [review]
Changes look good to me except for the codespell error. Like Bridget, I'll save approval til that's fixed.

File: test/e2e/velero_utils.go
Comment (nrb) [inline test/e2e/velero_utils.go:240]: I think this is fine for now, but I think eventually we should loosen this up. Those aren't optimal configurations for sure, but I think it would be useful to test AWS nodes/volumes with for example Azure blob storage.

File: test/e2e/velero_utils.go
Comment (nrb) [inline test/e2e/velero_utils.go:266]: I think we want to delete the CRDs too, as in https://velero.io/docs/v1.5/uninstalling/#docs

File: test/e2e/velero_utils.go
Comment (nrb) [inline test/e2e/velero_utils.go:270]: Could we set the velero namespace with `velero client config set namepace veleroNamepace` instead of passing it around all the time?

**Review by nrb — CHANGES_REQUESTED** [review]
The only blocking change I see here is to clean up CRDs on uninstall., or at least provide the option.

I can see leaving them in when testing upgrade scenarios, but to get a clean cluster, this second step is needed.

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go:240]: We will add that case in the future.

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go]: I added code to do that

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go:290]: We could, but this also allows us to test multiple veleros at the same time.  Let's revisit that later.

File: test/e2e/velero_utils.go
Comment (nrb) [inline test/e2e/velero_utils.go:275]: Thanks to #3447, we can re-use the definition of the label selector from the install package, so we only have 1 canonical place to change it. In particular, [this line](https://github.com/vmware-tanzu/velero/commit/6455350940eff352762c9b89e756fa23fe8add3b#diff-c6ec5e408c7f64e1ab2f86dcd1887b2486ee370c2142403a0f7c0a8f67349f5eR33) shows how to go from a map to a string.

**Review by nrb — CHANGES_REQUESTED** [review]
One very minor code reuse change, then I'm 👍

File: test/e2e/velero_utils.go
Comment (dsu-igeek) [reply test/e2e/velero_utils.go]: Done

**Comment by jenting** [comment]
The CI is failed at import ordering

File: test/e2e/README.md
Comment (jenting) [inline test/e2e/README.md:78]:
```suggestion
BSL_CONFIG="resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,storageAccount=$AZURE_STORAGE_ACCOUNT_ID,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID" BSL_BUCKET=<BUCKET_FOR_E2E_TEST_BACKUP> CREDS_FILE/path/to/aks-creds CLOUD_PROVIDER=azure make test-e2e
```

File: test/e2e/README.md
Comment (dsu-igeek) [reply test/e2e/README.md]: Yes, thanks.

File: test/e2e/README.md
Comment (jenting) [inline test/e2e/README.md:78]:
```suggestion
BSL_CONFIG="resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,storageAccount=$AZURE_STORAGE_ACCOUNT_ID,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID" BSL_BUCKET=<BUCKET_FOR_E2E_TEST_BACKUP> CREDS_FILE=/path/to/azure-creds CLOUD_PROVIDER=azure make test-e2e
```

**Review by jenting — COMMENTED** [review]
The rest LGTM

File: test/e2e/README.md
Comment (dsu-igeek) [reply test/e2e/README.md]: I updated this, no more README changes please.

**Review by zubron — APPROVED** [review]
Thanks, @dsu-igeek!

**Review by jenting — APPROVED** [review]
LGTM

**Pull request merged**
